controllers/util/secrets_util.go

```diff
@@ -0,0 +1,474 @@
+		controllerutil.RemoveFinalizer(secret, secretPolicyFinalizer)
+
+		if err := sutil.Client.Update(sutil.Ctx, secret); err != nil {
+			sutil.Log.Error(err, "unable to remove finalizer from secret", "secret", secret.Name)
+
+			return errorswrapper.Wrap(err, fmt.Sprintf("unable to remove finalizer from secret (secret: %s)",
+				secret.Name))
+		}
+	}
+
+	return nil
+}
+
+func inspectClusters(
+	clusters []plrv1.GenericClusterReference,
+	cluster string,
+	add bool) (bool, []plrv1.GenericClusterReference) {
+	found := true
+	survivors := []plrv1.GenericClusterReference{}
+
+	// Check if cluster is already part of placement rule
+	for _, plCluster := range clusters {
+		if plCluster.Name == cluster {
+			if add {
+				return found, clusters
+			}
+
+			continue
+		}
+		// Build a potential surviving cluster list
+		survivors = append(survivors, plCluster)
+	}
+
+	return !found, survivors
+}
+
+func (sutil *SecretsUtil) updatePlacementRule(
+	plRule *plrv1.PlacementRule,
+	secret *corev1.Secret,
+	cluster, namespace string,
+	add bool) error {
+	found, survivors := inspectClusters(plRule.Spec.Clusters, cluster, add)
+
+	switch add {
+	case true:
+		if found {
+			return nil
+		}
+
+		plRule.Spec.Clusters = append(plRule.Spec.Clusters, plrv1.GenericClusterReference{Name: cluster})
+	case false:
+		if len(survivors) == 0 {
+			return sutil.deletePolicyResources(secret, namespace)
+		}
+
+		if !found {
+			return nil
+		}
+
+		plRule.Spec.Clusters = survivors
+	}
+
+	err := sutil.Client.Update(sutil.Ctx, plRule)
+	if err != nil {
+		sutil.Log.Error(err, "unable to update placement rule", "placementRule", plRule.Name, "cluster", cluster)
+
+		return errorswrapper.Wrap(err, fmt.Sprintf("unable to update placement rule (placementRule: %s, cluster: %s)",
+			plRule.Name, cluster))
+	}
+
+	return nil
+}
+
+func (sutil *SecretsUtil) ticklePolicy(secret *corev1.Secret, namespace string) error {
+	policyName := fmt.Sprintf(secretResourceNameFormat, secretPolicyBaseName, secret.Name)
+	policyObject := gppv1.Policy{}
+
+	if err := sutil.Client.Get(sutil.Ctx,
+		types.NamespacedName{Namespace: namespace, Name: policyName},
+		&policyObject); err != nil {
+		sutil.Log.Error(err, "unable to get policy", "secret", secret.Name)
+
+		return errorswrapper.Wrap(err, fmt.Sprintf("unable to get policy (secret: %s)", secret.Name))
+	}
+
+	// Compare policy annotation to secret generation and trigger policy update if required
+	secretGenerationInPolicy := 0
+
+	for annotation, value := range policyObject.GetAnnotations() {
+		if annotation == policyTriggerAnnotation {
+			intValue, err := strconv.Atoi(value)
+			if err != nil {
+				sutil.Log.Error(err, "invalid policy trigger annotation value", "value", value)
+
+				return errorswrapper.Wrap(err, fmt.Sprintf("invalid policy trigger annotation value (value: %s)", value))
+			}
+
+			secretGenerationInPolicy = intValue
+
+			break
+		}
+	}
+
+	if secret.Generation == int64(secretGenerationInPolicy) {
```

Comment:
Argh! secrets do not have a `generation` number associated with them in the API server. Cannot use generation to trigger a policy update. May have to resort to incremental counter as suggested [here](https://github.com/stolostron/rhacm-docs/blob/2.4_stage/governance/custom_template.adoc#special-annotation-for-reprocessing) `A typical way to use this annotation is to increment the value by one each time.`

Reply:
Secrets do have a [`resourceVersion`](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes) though...it is used as a predicate for reconciling a drpolicy on secret and config map updates

Reply:
True, will think about this. Also, not sure why I am doing all the str->toi and backwards conversion, should be straightforward to just use the string version in one on the other.